src/WebJobs.Extensions.DurableTask/Listener/TaskEntityShim.cs

```diff
@@ -227,62 +237,90 @@ public override async Task<string> Execute(OrchestrationContext innerContext, st
+                    this.suspendAndContinueWithDelay = false;
+                    this.context.State.Suspended = true;
+                }
+
                 var jstate = JToken.FromObject(this.context.State);
 
                 // continue as new
                 innerContext.ContinueAsNew(jstate);
             }
 
-            if (this.context.ErrorsPresent(out var description))
-            {
-                this.Config.TraceHelper.FunctionFailed(
-                    this.context.HubName,
-                    this.context.Name,
-                    this.context.InstanceId,
-                    description,
-                    functionType: FunctionType.Entity,
-                    isReplay: false);
-            }
-            else
-            {
-                this.Config.TraceHelper.FunctionCompleted(
-                    this.context.HubName,
-                    this.context.Name,
-                    this.context.InstanceId,
-                    this.Config.GetIntputOutputTrace(this.context.State.EntityState),
-                    continuedAsNew: true,
-                    functionType: FunctionType.Entity,
-                    isReplay: false);
-            }
-
             // The return value is not used.
             return string.Empty;
         }
 
-        public async Task ExecuteBatch()
+        public async Task ExecuteBatch(CancellationToken onHostStopping)
         {
-            if (this.GetFunctionInfo().IsOutOfProc)
+            async Task ExecuteOperationsInBatch()
             {
-                if (this.operationBatch.Count > 0)
+                if (this.GetFunctionInfo().IsOutOfProc)
                 {
                     // process all operations in the batch using a single function call
                     await this.ExecuteOutOfProcBatch();
                 }
-            }
-            else
-            {
-                // call the function once per operation in the batch
-                for (this.BatchPosition = 0; this.BatchPosition < this.operationBatch.Count; this.BatchPosition++)
+                else
                 {
-                    var request = this.operationBatch[this.BatchPosition];
-                    await this.ProcessOperationRequestAsync(request);
+                    var stopwatch = new Stopwatch();
+                    stopwatch.Start();
+
+                    // call the function once per operation in the batch
+                    for (this.BatchPosition = 0; this.BatchPosition < this.operationBatch.Count; this.BatchPosition++)
+                    {
+                        // first, check if we should stop here and not execute the rest of the batch
+                        if (onHostStopping.IsCancellationRequested // host is shutting down
+                            || this.TimeoutTask.IsCompleted // we have timed out
+                            || stopwatch.Elapsed > TimeSpan.FromMinutes(1)) // we have spent significant time in this batch
```

Comment:
This 1-minute limitation might also be good to include in documentation since your changes could impact the throughput of existing entity code.

Reply:
yes, I will add it. 

I don't think it should reduce throughput significantly. The idea is that the interruption should be relatively brief (processing continues after saving state to disk and receiving the continuation signal) so I would think it gets amortized sufficiently if it happens once a minute only. Though I should probably test that.